src/virtual_dom/patch/patch_gen.rs

```diff
@@ -0,0 +1,540 @@
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum PatchKey {
+    Element {
+        namespace: Option<Namespace>,
+        tag: Tag,
+        el_key: Option<ElKey>,
+    },
+    Text,
+}
+
+fn make_key<Ms: 'static>(node: &Node<Ms>) -> Option<PatchKey> {
+    match node {
+        Node::Element(el) => Some(PatchKey::Element {
+            namespace: el.namespace.clone(),
+            tag: el.tag.clone(),
+            el_key: el.key.clone(),
+        }),
+        Node::Text(_) => Some(PatchKey::Text),
+        Node::Empty => None,
+    }
+}
+
+/// This is a command generator.
+/// See the module documenation for brief description of how this works.
+pub(crate) struct PatchGen<'a, Ms, OI, NI>
+where
+    Ms: 'static,
+    OI: Iterator<Item = Node<Ms>>,
+    NI: Iterator<Item = &'a mut Node<Ms>>,
+{
+    old_children_iter: Peekable<OI>,
+    new_children_iter: Peekable<NI>,
+    old_children: VecDeque<Node<Ms>>,
+    new_children: VecDeque<&'a mut Node<Ms>>,
+    matching_child_old: Option<OI::Item>,
+    matching_child_new: Option<NI::Item>,
+    matching_key: Option<PatchKey>,
+    keyed_mode: bool,
+}
+
+impl<'a, Ms, OI, NI> PatchGen<'a, Ms, OI, NI>
+where
+    Ms: 'static,
+    OI: Iterator<Item = Node<Ms>>,
+    NI: Iterator<Item = &'a mut Node<Ms>>,
+{
+    /// Creates the new `PatchGen` instance from the source iterators.
+    pub fn new(old_children_iter: OI, new_children_iter: NI) -> Self {
+        Self {
+            old_children_iter: old_children_iter.peekable(),
+            new_children_iter: new_children_iter.peekable(),
+            old_children: VecDeque::new(),
+            new_children: VecDeque::new(),
+            matching_child_old: None,
+            matching_child_new: None,
+            matching_key: None,
+            keyed_mode: false,
+        }
+    }
+
+    /// Consumes the source iterators and produces the patch command.
+    fn next_command(&mut self) -> Option<PatchCommand<'a, Ms>> {
+        if self.keyed_mode {
+            if self.matching_key.is_none()
+                && (self.old_children_iter.peek().is_some()
+                    || self.new_children_iter.peek().is_some())
+            {
+                self.matching_key = self.find_matching();
+            }
+            if self.matching_key.is_some() {
+                self.yield_keyed()
+            } else {
+                self.yield_remaining()
+            }
+        } else {
+            self.yield_keyless()
+        }
+    }
+
+    /// Takes children form the source iterators (new and old) and puts them in the
+    /// corresponding queues.
+    /// Stops when the key of the new child matches to any key of the previously seen old children
+    /// or the key of the old child matches to any key of the previously seen new children.
+    fn find_matching(&mut self) -> Option<PatchKey> {
+        // First store all seen keys to the sets.
+        // One for the old children.
+        let mut seen_old_keys: BTreeSet<_> = self
```

Comment:
What's your opinion about `BTreeSet` vs `HashSet` vs `FxHashSet`? I tried to find some best practices for using them some months ago but didn't find anything satisfying.
(It's just a question - I think you may have some experience with them as a "performance-oriented guy" ; feel free to close this comment.)

Reply:
I chose `BTreeSet` because in most cases set will have 1-10 items. I believe tree traversal  should have less overhead then hashing each item.
But I have not done any micro-benchmark.

Reply:
I share your opinion ; micro-benchmarks would be only killing time now, thanks!